File: kotlin/src/main/kotlin/com/dodopayments/checkout/DodoCheckout.kt
package com.dodopayments.checkout

import android.app.Activity
import android.content.Context
import androidx.activity.result.contract.ActivityResultContract
import kotlinx.coroutines.CompletableDeferred

/**
 * Entry point for the Dodo Payments mobile checkout.
 *
 * Open Dodo's hosted checkout in a Custom Tab and get a clean result from one
 * call:
 *
 * ```kotlin
 * val result = DodoCheckout.start(
 *     activity,
 *     CheckoutParams(
 *         checkoutUrl = checkoutUrl, // from your backend
 *         returnUrl = "myapp://checkout/return"
 *     )
 * )
 * ```
 *
 * The SDK contains **zero networking code** and holds **no API key**. It
 * watches navigation for your `return_url` and parses the result off the
 * query string. The result is a UI hint — grant access server-side from the
 * webhook.
 */
object DodoCheckout {

    /**
     * Presents Dodo's hosted checkout in a Custom Tab and resolves with the
     * outcome once the checkout navigates to `returnUrl` (or the user closes
     * the tab).
     *
     * @param activity The activity to launch the checkout screen from.
     * @param params Checkout URL and return URL.
     * @param onEvent Lifecycle callback for logging/analytics only — never
     *   decide the outcome from events. (Suspend-style only; the launcher-style
     *   [contract] path does not emit events.)
     * @return A [CheckoutResult] (UI hint — not proof of payment).
     * @throws CheckoutError for invalid input, a concurrent checkout, or a
     *   platform failure. A cancel or a declined payment is a *result*, not a
     *   thrown error.
     */
    suspend fun start(
        activity: Activity,
        params: CheckoutParams,
        onEvent: ((CheckoutEvent) -> Unit)? = null
    ): CheckoutResult {
        // Validate before touching any UI.
        UrlValidator.validateCheckoutUrl(params.checkoutUrl)
        UrlValidator.validateReturnUrl(params.returnUrl)
        RedirectResolution.ensureResolvable(activity, params.returnUrl)

        CheckoutCoordinator.guard.begin()

        val store = AbandonedSessionStore(SharedPreferencesKeyValueStore(activity))
        val deferred = CompletableDeferred<CheckoutResult>()
        CheckoutCoordinator.pendingResult = deferred
        CheckoutCoordinator.onEvent = onEvent

        try {
            // Record the session so it survives process death; the checkout
            // activity clears it on a clean finish.
            store.record(params.checkoutUrl)
            activity.startActivity(BrowserCheckoutHostActivity.newIntent(activity, params))
        } catch (t: Throwable) {
            CheckoutCoordinator.clearPending()
            CheckoutCoordinator.guard.end()
            store.clear()
            throw if (t is CheckoutError) {
                t
            } else {
                CheckoutError(
                    CheckoutError.Code.PLATFORM_ERROR,
                    t.message ?: "Failed to launch the checkout screen."
                )
            }
        }

        return try {
            deferred.await()
        } finally {
            CheckoutCoordinator.clearPending()
        }
    }

    /**
     * Launcher-style alternative to [start]:
     *
     * ```kotlin
     * val launcher = registerForActivityResult(DodoCheckout.contract()) { result -> … }
     * launcher.launch(CheckoutParams(checkoutUrl, returnUrl))
     * ```
     *
     * Same behavior. Validation errors throw [CheckoutError] out of
     * `launcher.launch(...)`. This path survives process death: if Android
     * kills your app during an external payment-app hop, the result is still
     * delivered to the re-registered callback.
     */
    fun contract(): ActivityResultContract<CheckoutParams, CheckoutResult> = CheckoutContract()

    /**
     * The session of a checkout that ended without a confirmed outcome, or
     * `null`.
     *
     * Set whenever the SDK never saw the return URL — the app was killed
     * mid-flow, or the result came back [CheckoutStatus.CANCELLED] because the
     * user dismissed the tab. Also set for [CheckoutStatus.PENDING], since
     * that status can itself mean an unparseable return URL rather than a
     * genuinely async payment method. Check it on launch *and* after every
     * CANCELLED or PENDING result, reconcile the session server-side, then
     * call [clearAbandonedSession].
     */
    fun getAbandonedSession(context: Context): AbandonedSession? =
        AbandonedSessionStore(SharedPreferencesKeyValueStore(context)).current()

    /** Clears the abandoned-session record after the merchant has reconciled it. */
    fun clearAbandonedSession(context: Context) {
        AbandonedSessionStore(SharedPreferencesKeyValueStore(context)).clear()
    }
}

File: flutter/lib/src/dodo_checkout.dart
import 'dart:async';

import 'package:flutter/foundation.dart' show visibleForTesting;
import 'package:flutter/services.dart' show PlatformException;

import 'checkout_exception.dart';
import 'checkout_types.dart';
import 'messages.g.dart';

/// Entry point for the Dodo Payments mobile checkout (Flutter).
///
/// This is a **thin wrapper**: all checkout logic lives in the native iOS
/// (`DodoCheckout`, Swift) and Android (`com.dodopayments:checkout-android`)
/// cores. The wrapper passes the call through over a typed Pigeon channel and
/// maps the result back.
///
/// ```dart
/// final result = await DodoCheckout.instance.start(
///   CheckoutParams(
///     checkoutUrl: Uri.parse(checkoutUrlFromYourBackend),
///     returnUrl: Uri.parse('myapp://checkout/return'),
///   ),
/// );
/// ```
///
/// The SDK contains **zero networking code** and holds **no API key**. The
/// result is a UI hint — grant access server-side from the webhook.
class DodoCheckout {
  DodoCheckout._(this._api) {
    DodoCheckoutFlutterApi.setUp(_FlutterApiImpl(_eventController));
  }

  /// Creates a facade over a custom host API — for tests only.
  @visibleForTesting
  DodoCheckout.forTesting(DodoCheckoutHostApi api) : _api = api;

  /// Simulates a native-pushed lifecycle event — for tests only.
  @visibleForTesting
  void debugEmitEvent(NativeCheckoutEvent event) => _eventController.add(event);

  /// The shared instance backed by the real platform channel.
  static final DodoCheckout instance = DodoCheckout._(DodoCheckoutHostApi());

  final DodoCheckoutHostApi _api;
  final StreamController<NativeCheckoutEvent> _eventController =
      StreamController<NativeCheckoutEvent>.broadcast();

  /// Presents Dodo's hosted checkout in a system browser tab
  /// (`SFSafariViewController` on iOS, a Custom Tab on Android) and completes
  /// with the outcome once the checkout navigates to `returnUrl` (or the user
  /// closes the tab).
  ///
  /// Throws a [CheckoutException] for invalid input, a concurrent checkout
  /// ([CheckoutErrorCode.alreadyInProgress] — only one checkout can run at a
  /// time), or a platform failure. A cancel or a declined payment is a
  /// *result*, not an exception.
  Future<CheckoutResult> start(CheckoutParams params) async {
    final void Function(CheckoutEvent event)? onEvent = params.onEvent;
    final StreamSubscription<NativeCheckoutEvent>? subscription = onEvent ==
            null
        ? null
        : _eventController.stream.listen((NativeCheckoutEvent event) {
            onEvent(CheckoutEvent(
              type: switch (event.type) {
                NativeEventType.opened => CheckoutEventType.opened,
                NativeEventType.returnReceived =>
                  CheckoutEventType.returnReceived,
                NativeEventType.closed => CheckoutEventType.closed,
              },
            ));
          });

    try {
      final NativeCheckoutResult native =
          await _api.start(_toStartRequest(params));
      return _toCheckoutResult(native);
    } on PlatformException catch (e) {
      throw CheckoutException(
        CheckoutException.codeFromNative(e.code),
        e.message ?? e.code,
      );
    } finally {
      await subscription?.cancel();
    }
  }

  /// The session of a checkout that ended without a confirmed outcome, or
  /// `null`.
  ///
  /// Set whenever the SDK never saw a return URL it could resolve to a durable
  /// outcome — the app was killed mid-flow, `start` completed with
  /// [CheckoutStatus.cancelled] because the user dismissed the browser, or it
  /// completed with [CheckoutStatus.pending], which is also the fallback for
  /// an unparseable return URL. Check this on launch *and* after every
  /// `cancelled` or `pending` result, reconcile the session server-side, then
  /// call [clearAbandonedSession] once the outcome is terminal.
  Future<AbandonedSession?> getAbandonedSession() async {
    final NativeAbandonedSession? native = await _api.getAbandonedSession();
    if (native == null) return null;
    return AbandonedSession(
      sessionId: native.sessionId,
      createdAt: DateTime.fromMillisecondsSinceEpoch(
        native.createdAtMillis,
        isUtc: true,
      ),
    );
  }

  /// Clears the abandoned-session record after it has been reconciled.
  Future<void> clearAbandonedSession() => _api.clearAbandonedSession();

  /// iOS only — forward incoming URLs here from your app's own URL-handling
  /// (e.g. an `app_links`/`uni_links` listener). Required: `SFSafariViewController`
  /// has no in-process way to catch its own return URL. A no-op returning
  /// `false` on Android. Returns `true` if the URL belonged to an in-flight
  /// checkout.
  Future<bool> handleOpenURL(String url) => _api.handleOpenURL(url);

  // -- Mapping between the public types and the Pigeon wire DTOs. -----------

  static StartRequest _toStartRequest(CheckoutParams params) {
    return StartRequest(
      checkoutUrl: params.checkoutUrl.toString(),
      returnUrl: params.returnUrl.toString(),
      customization: _toNativeBrowserCustomization(
        params.customization,
      ),
    );
  }

  static NativeBrowserCustomization? _toNativeBrowserCustomization(
    BrowserCustomization? customization,
  ) {
    if (customization == null) return null;
    final AndroidBrowserOptions? android = customization.android;
    final IosBrowserOptions? ios = customization.ios;
    return NativeBrowserCustomization(
      android: android == null
          ? null
          : NativeAndroidBrowserOptions(
              toolbarColor: android.toolbarColor?.toARGB32(),
              closeButtonStyle: switch (android.closeButtonStyle) {
                null => null,
                CloseButtonStyle.standard => NativeCloseButtonStyle.standard,
                CloseButtonStyle.back => NativeCloseButtonStyle.back,
              },
              closeButtonPosition: switch (android.closeButtonPosition) {
                null => null,
                CloseButtonPosition.start => NativeCloseButtonPosition.start,
                CloseButtonPosition.end => NativeCloseButtonPosition.end,
              },
              shareButtonEnabled: android.shareButtonEnabled,
              showTitleEnabled: android.showTitleEnabled,
              urlBarHidingEnabled: android.urlBarHidingEnabled,
              bookmarksButtonEnabled: android.bookmarksButtonEnabled,
              downloadsButtonEnabled: android.downloadsButtonEnabled,
              navigationBarColor: android.navigationBarColor?.toARGB32(),
              navigationBarDividerColor: android.navigationBarDividerColor
                  ?.toARGB32(),
              colorScheme: switch (android.colorScheme) {
                null => null,
                BrowserColorScheme.system => NativeBrowserColorScheme.system,
                BrowserColorScheme.light => NativeBrowserColorScheme.light,
                BrowserColorScheme.dark => NativeBrowserColorScheme.dark,
              },
            ),
      ios: ios == null
          ? null
          : NativeIosBrowserOptions(
              dismissButtonStyle: switch (ios.dismissButtonStyle) {
                null => null,
                DismissButtonStyle.done => NativeDismissButtonStyle.done,
                DismissButtonStyle.close => NativeDismissButtonStyle.close,
                DismissButtonStyle.cancel => NativeDismissButtonStyle.cancel,
              },
              barCollapsingEnabled: ios.barCollapsingEnabled,
              presentationStyle: switch (ios.presentationStyle) {
                null => null,
                PresentationStyle.pageSheet =>
                  NativePresentationStyle.pageSheet,
                PresentationStyle.fullScreen =>
                  NativePresentationStyle.fullScreen,
              },
              colorScheme: switch (ios.colorScheme) {
                null => null,
                BrowserColorScheme.system => NativeBrowserColorScheme.system,
                BrowserColorScheme.light => NativeBrowserColorScheme.light,
                BrowserColorScheme.dark => NativeBrowserColorScheme.dark,
              },
            ),
    );
  }

  static CheckoutResult _toCheckoutResult(NativeCheckoutResult native) {
    return CheckoutResult(
      status: switch (native.status) {
        NativeCheckoutStatus.succeeded => CheckoutStatus.succeeded,
        NativeCheckoutStatus.failed => CheckoutStatus.failed,
        NativeCheckoutStatus.cancelled => CheckoutStatus.cancelled,
        NativeCheckoutStatus.pending => CheckoutStatus.pending,
        NativeCheckoutStatus.expired => CheckoutStatus.expired,
      },
      paymentId: native.paymentId,
      subscriptionId: native.subscriptionId,
      licenseKeys: native.licenseKeys,
      customerEmail: native.customerEmail,
      raw: Map<String, String>.unmodifiable(native.raw),
    );
  }
}

/// Forwards native-pushed lifecycle events onto the broadcast stream that
/// [DodoCheckout.start] listens on for the duration of each call.
class _FlutterApiImpl implements DodoCheckoutFlutterApi {
  _FlutterApiImpl(this._controller);

  final StreamController<NativeCheckoutEvent> _controller;

  @override
  void onCheckoutEvent(NativeCheckoutEvent event) {
    _controller.add(event);
  }
}
